external-crates/move/crates/move-vm-runtime/src/interpreter/eval.rs

```diff
@@ -0,0 +1,1156 @@
+
+use crate::{
+    interpreter::state::{CallFrame,  MachineState, TypeWithLoader},
+    loader::{
+        arena::ArenaPointer,
+        ast::{Bytecode, Function},
+        Loader, ModuleDefinitionResolver,
+    },
+    native_functions::NativeContext,
+    native_extensions::NativeContextExtensions,
+    trace,
+};
+use fail::fail_point;
+use move_binary_format::{errors::*, file_format::JumpTableInner};
+use move_core_types::{
+    account_address::AccountAddress,
+    gas_algebra::{NumArgs, NumBytes},
+    vm_status::StatusCode,
+};
+use move_vm_config::runtime::{VMConfig, VMRuntimeLimitsConfig};
+#[cfg(feature = "gas-profiler")]
+use move_vm_profiler::GasProfiler;
+use move_vm_profiler::{
+    profile_close_frame, profile_close_instr, profile_open_frame, profile_open_instr,
+};
+use move_vm_types::{
+    data_store::DataStore,
+    gas::{GasMeter, SimpleInstruction},
+    loaded_data::runtime_types::Type,
+    values::{
+        IntegerValue, Reference, Struct, StructRef, VMValueCast, Value, Variant, VariantRef,
+        Vector, VectorRef,
+    },
+};
+use smallvec::SmallVec;
+
+use std::collections::VecDeque;
+
+const DEBUG_STEP_PRINT: bool = false;
+
+#[derive(PartialEq, Eq)]
+enum StepStatus {
+    Running,
+    Done,
+}
+
+struct RunContext<'loader, 'native, 'native_lifetimes> {
+    loader: &'loader Loader,
+    link_context: AccountAddress,
+    defn_resolver: ModuleDefinitionResolver<'loader>,
+    extensions: &'native mut NativeContextExtensions<'native_lifetimes>,
+}
+
+impl RunContext<'_, '_, '_> {
+    // TODO: The Run Context should hold this, not go get it from the Loader.
+    fn vm_config(&self) -> &VMConfig {
+        self.loader.vm_config()
+    }
+}
+
+macro_rules! set_err_info {
+    ($frame:expr, $e:expr) => {{
+        let function = $frame.function();
+        $e.at_code_offset(function.index(), $frame.pc)
+            .finish($frame.location())
+    }};
+}
+
+/// Main loop for the execution of a function.
+///
+/// This runs a newly-made Machine until it is complete. It expects the Machine to have a current
+/// call frame set up, with no operands on the stack and no existing call stack. It runs in a loop,
+/// calling `step`, until the call stack is empty.
+pub fn run(
+    start_state: MachineState,
+    data_store: &impl DataStore,
+    gas_meter: &mut impl GasMeter,
+    extensions: &mut NativeContextExtensions,
+    loader: &Loader,
+) -> VMResult<Vec<Value>> {
+    let link_context = data_store.link_context();
+    let resolver = start_state.current_frame.resolver(link_context, loader);
+
+    let mut run_context = RunContext {
+        loader,
+        link_context,
+        defn_resolver: resolver,
+        extensions,
+    };
+
+    let mut state = start_state;
+
+    if DEBUG_STEP_PRINT {
+        let mut buf = String::new();
+        let _ = state.debug_print_stack_trace(&mut buf, loader);
+        println!("Call Frame:\n{:?}", state.current_frame);
+        println!("{buf}");
+        buf.clear();
```

Comment:
this and next are not needed any longer if we put the `String` in the block right?